pyfr/plugins/tavg.py

```diff
@@ -166,44 +253,98 @@ def __call__(self, intg):
             currex = self._eval_acc_exprs(intg)
 
             # Accumulate them; always do this even when just writing
-            for a, p, c in zip(self.accex, self.prevex, currex):
-                a += 0.5*(intg.tcurr - self.prevt)*(p + c)
-
+            self._acc_avg_var(intg, currex, doaccum, dowrite)
+            
             # Save the time and solution
             self.prevt = intg.tcurr
             self.prevex = currex
 
             if dowrite:
                 comm, rank, root = get_comm_rank_root()
-
-                if self.mode == 'windowed':
-                    accex = self.accex
-                    tstart = self.tout_last
-                else:
-                    for a, c in zip(self.accex, self.caccex):
-                        c += a
-
-                    accex = self.caccex
-                    tstart = self.tstart_actual
-
+                accex = self.accex
+                vaccex = self.vaccex
+                tpts = self.tpts
+               
                 # Normalise the accumulated expressions
-                tavg = [a / (intg.tcurr - tstart) for a in accex]
-
-                # Evaluate any functional expressions
-                if self.fexprs:
-                    funex = self._eval_fun_exprs(intg, tavg)
+                tavg = [a / (2*(intg.tcurr - self.tstart_acc)) for a in accex]
+
+                # Calculate standard deviation
+                dev = [np.sqrt(np.abs(v / (2*(intg.tcurr - self.tstart_acc))))
+                        for v in vaccex]
+
+                # Maximum and sum of standard deviations
+                max_dev = np.zeros_like(self.anames, dtype=np.float64)
```

Comment:
Group this with the functional code as there is a lot of duplication.  So allocate a big array the length of anames + fnames.  Then just run through both together so we just have a single set of reductions rather than two.